Tally arithmetic

wbinventor opened the pull request:
This PR adds tally arithmetic to the Python API as initially proposed in #383. This is inspired by the tally arithmetic in thePINSPEC code, and the need to facilitate simple, vectorized uncertainty propagation for my upcoming multi-group cross-section module. In a nutshell, tally arithmetic allows two or more tallies and/or scalar values to be arithmetically combined into new "derived tallies." The goal is to facilitate rapid tally data processing with automated uncertainty propagation. 

The tally arithmetic's uncertainty propagation makes the assumption that the random variates represented by the combined tally bins are independent. Although this is almost never the case, it  for a simple and general first-order approximation to be made to the uncertainty propagation equations without the need to compute covariances with multiple samples of each variate (which would require multiple, independent simulations).

The following itemizes some of the new features available with the introduction of this PR.

**Arithmetic operations**
The following arithmetic operations can now be performed between two ``Tally`` objects, or a ``Tally`` object and a scalar constant:
* addition (*i.e.*, ``__add__``, ``__radd__``)
* subtraction (*i.e.*, ``__sub__``, ``__rsub__``)
* multiplication (*i.e.*, ``__mul__``, ``__rmul__``)
* division (*i.e.*, ``__div__``, ``__rdiv__``)
* exponentiation (*i.e.*, ``__exp__``)

**Tally outer products**
With tally arithmetic, it is necessary to somehow systematically map the filters, scores and nuclides between each ``Tally`` operand. This can be particularly problematic to do for ``Tally`` operands with different filters, scores and nuclides. The system I devised for this is to compute "outer products" of the filters bins, scores, and nuclides of each ``Tally`` operand, *if and only if* the filter bins, scores and nuclides in each operand do not match. As a result, the following new classes are introduced to represent the mapping between multiple filters, scores and nuclides:
* ``CrossScore`` - all enumerated paired combinations of two lists of scores
* ``CrossNuclide`` - all enumerated paired combinations of two lists of nuclides
* ``CrossFilter`` - all enumerated paired combintaions of two lists of filters

**Tally slicing**
One final new feature is tally slicing. This feature is inspired by NumPy array [*slicing*](http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html) (fancy indexing). The new ``Tally.slice(...)`` routine constructs a new derived ``Tally`` which encapsulates a subset of the data based on the user-defined filters, scores and nuclides input parameters. Slicing tallies can make the above tally arithmetic operations useful with finer granularity, especially when combined with the *tally merging* feature added in #387 to allow for optimized tally lookups under-the-hood at runtime.

This IPython Notebook ([link](https://gist.github.com/wbinventor/caa0286512c80da5f474)) illustrates the (very) basic use of tally arithmetic. The new features in this PR expose a great many degrees of freedom to the user for tally data processing. Although I have done my best to test them, it is without question that there are use cases I have not considered. Hopefully, other developers will enjoy poking as many holes into this as possible so we can really hammer things out.

paulromano commented:
Looks like the merge of #428 into develop caused conflicts. Can you merge develop to resolve those?

wbinventor commented:
The conflicts have been resolved.

paulromano commented on openmc/filter.py:
```diff
@@ -322,17 +323,17 @@ def get_bin_index(self, filter_bin):
             # Use lower energy bound to find index for energy Filters
             elif self.type in ['energy', 'energyout']:
-                val = self.bins.index(filter_bin[0])
+                val = np.where(self.bins == filter_bin[0])[0][0]
```
I'm assuming this change is to support cases where self.bins is a numpy array rather than a list?

wbinventor replied:
I changed the ``@bins.setter`` decorator such that ``bins`` is always a numpy array.

paulromano commented on openmc/tallies.py:
```diff
@@ -897,10 +904,18 @@ def get_values(self, scores=[], filters=[], filter_bins=[],
+                            self.get_filter_index(filter.type, bin))
+                    else:
+                        filter_indices[i].append(
+                            self.get_filter_index(filter.type, bin))
```
Confused about this if/else block since they appear to be the same

wbinventor replied:
Good point. They weren't the same at the in my initial implementation, but it appears they are now as I worked the algorithms out.

paulromano commented on openmc/tallies.py:
```diff
@@ -897,10 +904,18 @@ def get_values(self, scores=[], filters=[], filter_bins=[],
+                            self.get_filter_index(filter.type, bin))
+
+                # Account for stride in each of the previous filters
+                reverse_mult = lambda x: [xi * filter.num_bins for xi in x]
```
PEP8: don't assign a lambda to a named variable, use a def() instead

wbinventor replied:
Sure, the lambda isn't needed now anyway since I simplified to your proposed version below.

paulromano commented on openmc/tallies.py:
```diff
@@ -897,10 +904,18 @@ def get_values(self, scores=[], filters=[], filter_bins=[],
+
+                # Account for stride in each of the previous filters
+                reverse_mult = lambda x: [xi * filter.num_bins for xi in x]
+                filter_indices[:i] = map(reverse_mult, filter_indices[:i])
```
Isn't `filter_indices[:i] *= filter.num_bins` equivalent to what you have here?

wbinventor replied:
No since ``filter_indices`` is a list rather than a numpy array. Doing this will simply increase the length of the list with duplicate elements, whereas it would . However, since you raised this point I found a way to make ``filter_indices`` a numpy array which simplifies things a bit. Thanks for pointing this out.

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+
+        Parameters
+        ----------
+        other : Tally or Integer or Rational
```
Should Rational be Real instead? Consider the following:
```
In [1]: from numbers import Rational, Real

In [2]: isinstance(1, Rational)
Out[2]: True

In [3]: isinstance(1.5, Rational)
Out[3]: False

In [4]: isinstance(1.5, Real)
Out[4]: True
```

wbinventor replied:
sure

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            self._outer_product(other, new_tally, binary_op='+')
+            data = self._align_tally_data(other)
+            new_tally._mean = data['self']['mean'] + data['other']['mean']
+            new_tally._std_dev = np.sqrt(data['self']['std. dev.']**2 + \
```
No need for backslash

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            self._outer_product(other, new_tally, binary_op='-')
+            data = self._align_tally_data(other)
+            new_tally._mean = data['self']['mean'] - data['other']['mean']
+            new_tally._std_dev = np.sqrt(data['self']['std. dev.']**2 + \
```
No need for backslash

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            The tally on the right hand side of the outer product
+        new_tally: Tally
+            The new tally to represent the outer product
+        op : str
```
binary_op rather than op

paulromano replied:
The numpydoc syntax for a limited set of choices would be:
`binary_op : {'+', '-', '*', '/', '^'}`

wbinventor replied:
updated with the numpydoc syntax

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+                  'since it does not contain any results.'.format(self.id)
+            raise ValueError(msg)
+
+        new_tally = Tally(name='derived')
```
Why is `name='derived'` only given in `__div__` and `__pow__` but not the others?

wbinventor replied:
Simple mistake. I am going to make them all 'derived' unless someone objects.

paulromano commented:
A design question: you introduce two methods _outer_product and _align_tally_data. It seems that _align_tally_data always follows _outer_product and then is used to set the mean and standard deviation of the new tally. What do you think of having the _outer_product method call the _align_tally_data method itself, set the mean and standard deviation appropriately, and return a new instance of a Tally. It would make the method appear more operator-like. So, e.g., the following chunk of code:

```python
new_tally = Tally()
new_tally.with_batch_statistics = True

if isinstance(other, Tally):

    # Check that results have been read                                                                                                                       
    if other.sum is None:
        msg = 'Unable to use tally arithmetic with Tally ID={0} ' \
              'since it does not contain any results.'.format(other.id)
        raise ValueError(msg)

    self._outer_product(other, new_tally, binary_op='+')
    data = self._align_tally_data(other)
    new_tally._mean = data['self']['mean'] + data['other']['mean']
    new_tally._std_dev = np.sqrt(data['self']['std. dev.']**2 + \
                                 data['other']['std. dev.']**2)
```

would become:
```python
if isinstance(other, Tally):

    # Check that results have been read                                                                                                                       
    if other.sum is None:
        msg = 'Unable to use tally arithmetic with Tally ID={0} ' \
              'since it does not contain any results.'.format(other.id)
        raise ValueError(msg)

    new_tally = self._outer_product(other, binary_op='+')
```

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+                new_tally.add_filter(self_filter)
+        else:
+            all_filters = [self.filters, other.filters]
+            for self_filter, other_filter in itertools.product(*all_filters):
```
Why create a list with two items and unpack it rather than just passing two arguments to itertools.product?

wbinventor replied:
The only reason was to make the ``for`` loop cleanly fit on a single line with 80 chars.

paulromano replied:
Ah ok. That's fine with me.

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+                new_tally.add_filter(new_filter)
+
+        # Generate score "outer products"
+        if self.scores == other.scores:
```
Should this be `np.all(self.scores == other.scores)`?

paulromano replied:
Oh nevermind, I just thought about this again. The simple equality should be ok.

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            other_num_filter_bins = other.num_filter_bins
+            num_filter_bins = self_num_filter_bins * other_num_filter_bins
+            self_repeat_factor = num_filter_bins / self_num_filter_bins
+            other_tile_factor = num_filter_bins / other_num_filter_bins
```
Isn't this equivalent to:
```
self_repeat_factor = other_num_filter_bins
other_tile_factor = self_num_filter_bins
```

wbinventor replied:
Haha, yes indeed! This code was a mess at first as I tried different ideas out, and I didn't realize that it had simplified to this point.

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            new_tally._std_dev = np.sqrt(data['self']['std. dev.']**2 + \
+                                         data['other']['std. dev.']**2)
+
+        elif isinstance(other, (Integral, Rational)):
```
Whether Rational or Real is the right choice, you shouldn't need the check against Integral since integers are reals/rationals.

wbinventor replied:
right of course

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            new_tally._std_dev = np.sqrt(data['self']['std. dev.']**2 + \
+                                         data['other']['std. dev.']**2)
+
+        elif isinstance(other, (Integral, Rational)):
```
Integers are reals/rationals

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            new_tally._std_dev = np.abs(new_tally.mean) * \
+                                 np.sqrt(self_rel_err**2 + other_rel_err**2)
+
+        elif isinstance(other, (Integral, Rational)):
```
Integers are reals/rationals

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            new_tally._std_dev = np.abs(new_tally.mean) * \
+                                 np.sqrt(self_rel_err**2 + other_rel_err**2)
+
+        elif isinstance(other, (Integral, Rational)):
```
Integers are reals/rationals

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+            new_tally._std_dev = np.abs(new_tally.mean) * \
+                                 np.sqrt(first_term**2 + second_term**2)
+
+        elif isinstance(power, (Integral, Rational)):
```
Integers are reals/rationals

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+
+        """
+
+        return self * (1. / other)
```
We really want other/self, not self/other, no? I suppose you could do `other * self**-1`.

wbinventor replied:
Yes, I agree and like your solution

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+                    score_indices.append(score_index)
+
+            # Loop over indices in reverse to remove excluded scores
+            for score_index in score_indices[::-1]:
```
Could you also use `reversed(score_indices)` here?

wbinventor replied:
Sure, is there a reason to?

paulromano replied:
`[::-1]` creates a shallow copy whereas `reversed()` does not

paulromano commented on openmc/tallies.py:
```diff
@@ -1410,6 +1432,765 @@ def export_results(self, filename='tally-results', directory='.',
+
+        Parameters
+        ----------
+        other : Tally or Integer or Rational
```
power :

paulromano commented:
When I try to run the ipython notebook you posted, it fails at the first call to get_pandas_dataframe because in get_pandas_dataframe it checks whether self._sum and self._sum_sq are None for the derived tally (which they are). Perhaps the check should be on mean and std_dev instead?

paulromano commented:
@wbinventor Another really awesome PR! I think my review is pretty much complete. Happy to merge once you resolve all the comments.

wbinventor commented:
Yes you are right. I actually switched back to sum/sum_sq this morning when
fixing the merge conflicts with develop since I double guessed myself.
On Aug 5, 2015 7:17 PM, "Paul Romano" <notifications@github.com> wrote:

> When I try to run the ipython notebook you posted, it fails at the first
> call to get_pandas_dataframe because in get_pandas_dataframe it checks
> whether self._sum and self._sum_sq are None for the derived tally (which
> they are). Perhaps the check should be on mean and std_dev instead?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/426#issuecomment-128215774>.
>

wbinventor commented:
@paulromano I've revised this PR per your comments. I refactored the ``Tally._outer_product(...)`` and operator overloading routines to be more "operator-like" as you suggested. In the process, I also added a new ``_derived`` class attribute to ``Tally``. This new attribute allows for better error checking in the tally arithmetic routines. In particular, these routines check that the ``_sum`` attribute for each ``Tally`` is populated with data. However, if a user were to want to do tally arithmetic with a derived ``Tally`` - *i.e.*, a ``Tally`` created by one of the overloaded operators - its ``_sum`` attribute would be ``None`` since there is not correct way to compute a derived version of this. Hence, the error checking now uses a combination of ``_derived`` and ``_sum`` now.

paulromano commented:
Looks great now! Ready to merge pending resolution of the `[::-1]` vs `reversed` comment

wbinventor commented:
I revised to use ``reversed(...)`` over score and bin indices.
